luigi/contrib/nomad.py

```diff
@@ -0,0 +1,347 @@
+            self.__logger.info("Fetching logs from " + allocation["Name"])
+            if len(logs) > 0:
+                for line in logs.split('\n'):
+                    self.__logger.info(line)
+
+    def __print_nomad_hints(self):
+        self.__logger.info("To stream allocation logs, use:")
+        self.__logger.info("`nomad logs -f -job %s`" % (self.uu_name))
+
+    def __verify_job_has_started(self):
+        """Asserts that the job has successfully started"""
+        # Verify that the job started
+        job = self.__get_job()
+
+        # Verify that allocation exists
+        allocations = self.__get_allocations()
+        if not allocations:
+            self.__logger.debug(
+                'No allocations found for %s, waiting for cluster state to match the job definition' % self.uu_name
+            )
+            time.sleep(self.allocation_wait_interal)
+            allocations = self.__get_allocations()
+
+        assert len(allocations) > 0, "No allocations scheduled by " + self.uu_name
+        if job["Status"] == "pending":
+            return False
+
+        return True
+
+    def __get_job_status(self):
+        """Return the Nomad job status"""
+        # Figure out status and return it
+        job = self.__get_job()
+
+        if job["Status"] == "dead":
+            allocations = self.__nomad.job.get_allocations(self.uu_name)
+            for allocation in allocations:
+                if allocation["ClientStatus"] == "complete":
+                    if self.print_allocation_logs_on_exit:
+                        self.__print_allocation_logs()
+                    if self.delete_on_success:
+                        self.__nomad.job.deregister_job(self.uu_name)
+                    return "SUCCEEDED"
+                elif allocation["ClientStatus"] == "failed":
+                    self.__logger.debug("Nomad job " + self.uu_name + " failed")
+                    if self.print_allocation_logs_on_exit:
+                        self.__print_allocation_logs()
+                    return "FAILED"
+        return "RUNNING"
+
+    def run(self):
+        self._init_nomad()
+        # Render job
+        job_json = {"Job": self.spec_schema}
+        job_json["Job"]["Id"] = self.uu_name
+        labels = {
+            "spawned_by": "luigi",
+            "luigi_task_id": self.job_uuid
+        }
+        if "Meta" not in job_json["Job"] or job_json['Job']['Meta'] is None:
+            job_json['Job']['Meta'] = labels
+        else:
+            job_json['Job']['Meta'].update(labels)
+        if self.nomad_namespace is not None:
+            job_json['Namespace'] = self.nomad_namespace
+        if "Datacenters" not in job_json["Job"]:
+            job_json["Job"]["Datacenters"] = ["dc1"]
+        if "TaskGroups" in job_json["Job"]:
+            for tg in job_json["Job"]["TaskGroups"]:
+                if "RestartPolicy" not in tg:
+                    tg["RestartPolicy"] = {
```

Comment:
I divided it up a bit further, that should hopefully make it more self-explanatory.

It looks a bit complex just because that it respects any user spec, not overriding.